File: lcr/183/main.py
from typing import List
import collections


class Solution:
    def maxAltitude(self, heights: List[int], limit: int) -> List[int]:
        if not heights or limit == 0:
            return []

        deque = collections.deque()
        for i in range(limit):
            while deque and heights[i] > deque[-1]:
                deque.pop()
            deque.append(heights[i])

        results = [deque[0]]
        for i in range(limit, len(heights)):
            if heights[i-limit] == deque[0]:
                deque.popleft()

            while deque and heights[i] > deque[-1]:
                deque.pop()
            
            deque.append(heights[i])
            results.append(deque[0])
        
        return results

File: lcr/183/main.go
package main

func maxAltitude(heights []int, limit int) []int {
	if len(heights) == 0 && limit == 0 {
		return []int{}
	}
	deque := []int{}
	var i int
	for ; i < limit; i++ {
		for len(deque) > 0 && heights[i] > deque[len(deque)-1] {
			deque = deque[:len(deque)-1]
		}
		deque = append(deque, heights[i])
	}
	results := []int{deque[0]}
	for ; i < len(heights); i++ {
		if heights[i-limit] == deque[0] {
			deque = deque[1:]
		}

		for len(deque) > 0 && heights[i] > deque[len(deque)-1] {
			deque = deque[:len(deque)-1]
		}

		deque = append(deque, heights[i])
		results = append(results, deque[0])
	}

	return results
}

func main() {

}
